service/hproxy/hproxy.go

```diff
@@ -0,0 +1,960 @@
+		s.persistentConnections = connections
+		s.mtx.Unlock()
+
+		if s.promPollVerbose {
+			log.Infof("new: %v healthy: %v unhealthy: %v removed: %v "+
+				"persistent connections: %v", h.NewNodes,
+				h.HealthyNodes, h.UnhealthyNodes, h.RemovedNodes,
+				connections)
+		}
+		ticker.Reset(promPollFrequency)
+	}
+}
+
+func (s *Server) running() bool {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	return s.isRunning
+}
+
+func (s *Server) testAndSetRunning(b bool) bool {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	old := s.isRunning
+	s.isRunning = b
+	return old != s.isRunning
+}
+
+func (s *Server) promRunning() float64 {
+	r := s.running()
+	if r {
+		return 1
+	}
+	return 0
+}
+
+func (s *Server) isHealthy(_ context.Context) bool {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	for k := range s.hvmHandlers {
+		// If we see a single hvm unhealthy bail with unhealthy status.
+		// XXX i think this is technically correct but we do return
+		// 503 to the caller here.
+		if s.hvmHandlers[k].state == StateUnhealthy {
+			return false
+		}
+	}
+	return true
+}
+
+func (s *Server) health(ctx context.Context) (bool, any, error) {
+	log.Tracef("health")
+	defer log.Tracef("health exit")
+
+	s.mtx.RLock()
+	h := s.promHealth
+	s.mtx.RUnlock()
+
+	return s.isHealthy(ctx), h, nil
+}
+
+type HVMState int
+
+type health struct {
+	NewNodes       int `json:"new_nodes"`
+	HealthyNodes   int `json:"healthy_nodes"`
+	UnhealthyNodes int `json:"unhealthy_nodes"`
+	RemovedNodes   int `json:"removed_nodes"`
+}
+
+const (
+	StateNew       HVMState = 0
+	StateHealthy   HVMState = 1
+	StateUnhealthy HVMState = 2
+	StateRemoved   HVMState = 3
+)
+
+var stateString = map[HVMState]string{
+	StateNew:       "new",
+	StateHealthy:   "healthy",
+	StateUnhealthy: "unhealthy",
+	StateRemoved:   "removed",
+}
+
+type HVMHandler struct {
+	id int // node id
+
+	rp *httputil.ReverseProxy
+	u  *url.URL     // Connection URL for HVM
+	c  *http.Client // In here to reuse connection
+
+	connections uint // Open connections
+
+	poking bool // true when getting health
+	poker  Proxy
+	state  HVMState // Do NOT directly set, use utility functions!
+}
+
+func (s *Server) handleProxyRequest(w http.ResponseWriter, r *http.Request) {
+	log.Tracef("handleProxyRequest: %v", r.RemoteAddr)
+	defer log.Tracef("handleProxyRequest exit: %v", r.RemoteAddr)
+
+	// Select host to call
+	s.mtx.Lock()
+	id := -1
+	if v, ok := s.clients[r.RemoteAddr]; ok {
+		// Only call same host when healthy
+		if s.hvmHandlers[v.node].state == StateHealthy {
+			id = v.node
+			v.reset(s.cfg.ClientIdleTimeout) // reset timer for reuse here
+		} else {
+			// hvm died, remove persisted client
+			s._clientRemove(r.RemoteAddr)
+		}
+	}
+
+	// Find suitable candidate
+	if id == -1 {
+		var leastConnections uint = math.MaxUint
+		for k := range s.hvmHandlers {
+			if s.hvmHandlers[k].state == StateHealthy {
+				if s.hvmHandlers[k].connections < leastConnections {
+					leastConnections = s.hvmHandlers[k].connections
+					id = k
```

Comment:
nit: we should `break` here when we find a healthy node, so we don't continue iterating

Reply:
We don’t know if the node we find first is the indeed the node with the least amount of connections. We must finish the loop, only shortcut is finding 0. Also, there won’t ever be more than like 5 VMs to look at.

Reply:
ah I see. makes sense.

also it's "the least _number_ of connections"